pkg/metrics_store/metrics_store.go

```diff
@@ -0,0 +1,110 @@
+package metricsstore
+
+import (
+	"sync"
+
+	"k8s.io/kube-state-metrics/pkg/metrics"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+)
+
+// MetricsStore implements the k8s.io/kubernetes/client-go/tools/cache.Store
+// interface.
+type MetricsStore struct {
+	mutex   sync.RWMutex
+	metrics map[string][]*metrics.Metric
+
+	generateMetricsFunc func(interface{}) []*metrics.Metric
+}
+
+// NewMetricsStore returns a new MetricsStore
+func NewMetricsStore(generateFunc func(interface{}) []*metrics.Metric) *MetricsStore {
+	return &MetricsStore{
+		generateMetricsFunc: generateFunc,
+		metrics:             map[string][]*metrics.Metric{},
+	}
+}
+
+// Implementing k8s.io/kubernetes/client-go/tools/cache.Store interface
+
+func (s *MetricsStore) Add(obj interface{}) error {
+	o, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
```

Comment:
We should take some lock profiles of this on a large cluster, this looks like there is going to be some contention.